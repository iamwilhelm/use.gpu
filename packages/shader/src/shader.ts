import { Tree, SyntaxNode } from '@lezer/common';
import { parser } from './grammar/glsl';

export const parseGLSL = (code: string): Tree => {
  const parsed = parser.parse(code);
  return parsed;
}
