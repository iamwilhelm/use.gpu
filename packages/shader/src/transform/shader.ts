import { ShaderLanguages, ShaderCompiler } from '@use-gpu/core/types';
import { ParsedModule, ParsedModuleCache, ShaderDefine } from '../types';
import { Tree, SyntaxNode } from '@lezer/common';

import { makeASTParser, compressAST, decompressAST, getProgramHash } from './ast';
import { parser } from '../grammar/glsl';
import LRU from 'lru-cache';

type LangDef = {
  glsl?: ShaderCompiler,
  cache?: any,
};

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const loadModule = (
  code: string,
  name: string,
  compressed: boolean = false,
): ParsedModule => {
  if (code == null) throw new Error(`Shader code ${name} undefined`);
  let tree = parseGLSL(code);

  const parser = makeASTParser(code, tree);
  const table = parser.getSymbolTable();
  const shake = parser.getShakeTable(table);

  if (compressed) tree = decompressAST(compressAST(tree));

  return {name, code, tree, table, shake};
}

// Use cache to load modules
export const loadModuleWithCache = (
  code: string,
  name: string,
  cache: ParsedModuleCache | null = null,
): ParsedModule => {
  if (!cache) return loadModule(code, name, true);

  const hash = getProgramHash(code);
  const entry = cache.get(hash);
  if (entry) return entry;
  
  const module = loadModule(code, name, true);
  cache.set(hash, module);
  return module;
}

// Parse GLSL using lezer grammar
export const parseGLSL = (code: string): Tree => parser.parse(code);

// Make GLSL definitions
export const defineGLSL = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

// Make shader languages interface
export const makeShaderLanguages = (langs: LangDef[]): ShaderLanguages => {
  const out = {} as any;
  for (const {glsl} of langs) {
    if (glsl) {
      out.glsl = {
        compile: (code: string, stage: any) => (glsl as any)(code, stage, false),
        cache: makeModuleCache(),
      };
    }
  }
  return out;
};

// LRU cache for parsed shader code
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU({
  max: 100,
  ...options,
});
