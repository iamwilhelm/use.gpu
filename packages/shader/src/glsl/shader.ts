import { ParsedModule, ParsedModuleCache, ShaderDefine, ShaderCompiler, SymbolTable, VirtualTable } from '../types';
import { Tree, SyntaxNode } from '@lezer/common';

import { makeASTParser, compressAST, decompressAST } from './ast';
import { getProgramHash, makeKey } from '../util/hash';
import { parser } from '../grammar/glsl';
import { PREFIX_VIRTUAL } from '../constants';
import LRU from 'lru-cache';

const EMPTY_LIST = [] as any[];
const EMPTY_TABLE = {} as any;

// Parse a code module into its in-memory representation
// (AST + symbol table)
export const loadModule = (
  code: string,
  name: string,
  entry?: string,
  compressed: boolean = false,
): ParsedModule => {
  if (code == null) throw new Error(`Shader code ${name} undefined`);
  if (typeof code !== 'string') throw new Error(`Shader code ${name} is not a string`);
  let tree = parseShader(code);

  const parser = makeASTParser(code, tree);
  const table = parser.getSymbolTable();
  const shake = parser.getShakeTable(table);

  if (compressed) tree = decompressAST(compressAST(tree));

  return {name, code, table, entry, shake, tree};
}

// Use cache to load modules
export const loadModuleWithCache = (
  code: string,
  name: string,
  entry?: string,
  cache: ParsedModuleCache | null = DEFAULT_CACHE,
): ParsedModule => {
  if (!cache) return loadModule(code, name, entry, true);

  const hash = getProgramHash(code);
  const cached = cache.get(hash);
  if (cached) return {...cached, entry};
  
  const module = loadModule(code, name, entry, true);
  cache.set(hash, module);
  return {...module, entry};
}

// Load a static (inert) module
export const loadStaticModule = (code: string, name: string, entry?: string) => {
  const table = {
    hash: getProgramHash(code),
  };
  return { name, code, table, entry };
}

// Load a virtual (generated) module
export const loadVirtualModule = (
  virtual: VirtualTable,
  initTable: Partial<SymbolTable> = EMPTY_TABLE,
  entry?: string,
  key: string | number = makeKey(),
) => {
  let symbols = initTable.symbols ?? EMPTY_LIST;
  const code = `#virtual [${symbols.join(' ')}] ${key.toString(16)}`;

  const hash = getProgramHash(code);
  const name = `${PREFIX_VIRTUAL}${hash.slice(0, 6)}_`;

  const table = {
    hash,
    symbols,
    visibles: symbols,
    ...initTable,
  };
  return { name, code, table, entry, virtual };
}

// Parse GLSL using lezer grammar
export const parseShader = (code: string): Tree => parser.parse(code);

// Make GLSL definitions
export const defineConstants = (defs: Record<string, ShaderDefine>): string => {
  const out = [];
  for (let k in defs) if (defs[k] !== false && defs[k] !== null) out.push(`#define ${k} ${defs[k]}`);
  return out.join("\n");
}

// Make shader languages interface
export const makeLanguage = (compile: any, cache?: any) => ({
  compile: (code: string, stage: any) => (compile as any)(code, stage, false),
  cache: cache ?? makeModuleCache(),
});

// LRU cache for parsed shader code
export const makeModuleCache = (options: Record<string, any> = {}) => new LRU<string, ParsedModule>({
  max: 100,
  ...options,
});

export const DEFAULT_CACHE = makeModuleCache();
