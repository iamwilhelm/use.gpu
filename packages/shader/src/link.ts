import { Tree } from '@lezer/common';
import { SymbolTable } from './types';
import { parseGLSL } from './shader';
import { makeASTParser } from './ast';

type Program = {
  entry: null,
  code: string[],
};

type Module = {
  name: string,
  code: string,
  tree: Tree,
  table: SymbolTable,
};

export const makeProgram = () => ({});

export const linkModule = (code: string, library: Record<string, string>) => {

  const modules = loadModules(code, library);
  const namespaces = new Map<string, string>();
  const program = [] as string[];

  for (const module of modules) {
    const {name, code, tree, table} = module;
    const {hash, symbols, modules} = table;

    const namespace = '_' + table.hash.slice(0, 4) + '_';
    namespaces.set(name, namespace);

    const rename = new Map<string, string>();
    if (name !== 'main') {
      for (const {name} of symbols) {
        rename.set(name, namespace + name);
      }
    }
    
    for (const {name, imports} of modules) {
      const namespace = namespaces.get(name);
      for (const {name, imported} of imports) {
        rename.set(name, namespace + imported);
      }
    }
    
    console.log('module', name, namespace, Array.from(rename.values()));
    program.push(rewriteAST(code, tree, rename));
  }

  return program.join("\n");
}

export const rewriteAST = (code: string, tree: Tree, rename: Map<string, string>) => {
  return code;
}

export const loadModules = (code: string, library: Record<string, string>) => {
  const seen = new Map<string, number>();
  const out = [];

  const queue = [{name: 'main', code, depth: 0}];
  seen.set('main', 0);

  while (queue.length) {
    const {name, code, depth} = queue.shift()!;

    const tree = parseGLSL(code);
    const table = makeASTParser(code, tree).extractSymbolTable();

    out.push({name, code, tree, table});

    const {modules} = table;
    for (const {name} of modules) {      
      const code = library[name];
      if (!code) throw new Error(`Unknown module '${name}'`);
      
      if (!seen.has(name)) queue.push({name, code, depth: depth + 1});
      seen.set(name, depth + 1);
    }
  }

  out.sort((a, b) => seen.get(b.name)! - seen.get(a.name)! || a.name.localeCompare(b.name));

  return out;
}
