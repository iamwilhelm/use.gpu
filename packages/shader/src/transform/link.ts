import { Tree } from '@lezer/common';
import { SymbolTable } from '../types';
import { parseGLSL } from './shader';
import { makeASTParser, rewriteAST } from './ast';
import { GLSL_VERSION } from '../constants';
import * as T from '../grammar/glsl.terms';

type Program = {
  entry: null,
  code: string[],
};

type Module = {
  name: string,
  code: string,
  tree: Tree,
  table: SymbolTable,
};

export const makeProgram = () => ({});

export const linkModule = (
  code: string,
  library: Record<string, string>,
  cache: Map<string, Module> | null = null,
) => {

  const modules = loadModules(code, library);

  const program = [`#version ${GLSL_VERSION}`] as string[];
  const namespaces = new Map<string, string>();

  const used = new Set<string>();
  const exported = new Set<string>();

  for (const module of modules) {
    const {name, code, tree, table} = module;
    const {symbols, exports: exp, modules} = table;

    const namespace = reserveNamespace(module, namespaces, used);

    const rename = new Map<string, string>();
    if (name !== 'main') {
      for (const {name} of symbols) {
        rename.set(name, namespace + name);
        used.add(namespace + name);
      }
      for (const {name} of exp) {
        exported.add(namespace + name);
      }
    }

    for (const {name: module, imports} of modules) {
      const namespace = namespaces.get(module);
      for (const {name, imported} of imports) {
        const imp = namespace + imported;
        if (!used.has(imp)) console.warn(`Import ${name} from '${module}' does not exist`);
        else if (!exported.has(imp)) console.warn(`Import ${name} from '${module}' is private`);
        rename.set(name, imp);
      }
    }

    program.push(rewriteAST(code, tree, rename));
  }

  return program.join("\n");
}

export const loadModules = (code: string, library: Record<string, string>) => {
  const seen = new Map<string, number>();
  const out = [];

  const queue = [{name: 'main', code, depth: 0}];
  seen.set('main', 0);

  while (queue.length) {
    const {name, code, depth} = queue.shift()!;

    const tree = parseGLSL(code);
    const table = makeASTParser(code, tree).extractSymbolTable();

    out.push({name, code, tree, table});

    const {modules} = table;
    for (const {name} of modules) {      
      const code = library[name];
      if (!code) throw new Error(`Unknown module '${name}'`);
      
      if (!seen.has(name)) queue.push({name, code, depth: depth + 1});
      seen.set(name, depth + 1);
    }
  }

  out.sort((a, b) => seen.get(b.name)! - seen.get(a.name)! || a.name.localeCompare(b.name));

  return out;
}

export const reserveNamespace = (
  module: Module,
  namespaces: Map<string, string>,
  used: Set<string>,
) => {
  const {name, table: {hash}} = module;
  let namespace;
  let n = 2;

  do {
    namespace = '_' + hash.slice(0, n++) + '_';
  } while (used.has(namespace));

  namespaces.set(name, namespace);
  used.add(namespace);

  return namespace;
}
