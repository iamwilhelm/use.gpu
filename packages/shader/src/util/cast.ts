import { UniformAttribute, ShaderModule, ParsedBundle } from '../types';
import { loadVirtualModule } from '../util/shader';
import { getProgramHash } from '../util/hash';
import { toBundle } from '../util/bundle';
import { PREFIX_CAST } from '../constants';

const NO_SYMBOLS = [] as string[];

export type MakeCastAccessor = (
  name: string,
  accessor: string,
  args: string[],
  from: string,
  to: string,
  swizzle: string,
) => string;

export const toTypeString = (t: Type): string => {
  if (typeof t === 'object') {
    if (t.type) return toTypeString(t.type);
    if (t.args) return `${t.name}<${t.args.map(t => toTypeString(t)).join(',')}>`;
    else return t.name;
  }
  return t;
}

export const makeCastTo = (
  makeCastAccessor: MakeCastAccessor,
) => (
  bundle: ShaderModule,
  type: string,
  swizzle: string,
): ParsedBundle => {
  const {module, virtual} = toBundle(bundle);
  const {name, format, args} = bundleToAttribute(bundle);

  const {table: {hash}} = module;
  const rehash = getProgramHash(`#cast [${name} ${format}] ${hash}`);
  const symbols = ['cast', 'getValue'];
  const namespace = `${PREFIX_CAST}${rehash.slice(0, 6)}`;

  const imported = {
    at: 0,
    symbols: NO_SYMBOLS,
    name: namespace,
    imports: [{name: 'getValue', imported: name}],
  };
  const modules = [imported];

  // Code generator
  const render = (namespace: string, rename: Map<string, string>) => {
    const name = rename.get('cast') ?? 'cast';
    const accessor = rename.get('getValue') ?? 'getValue';
    return makeCastAccessor(name, accessor, args ?? [], format, type, swizzle);
  }

  const cast = loadVirtualModule(
    { render },
    { symbols, modules },
    'cast',
    rehash,
  );

  const revirtual = module.virtual
    ? (virtual ? [...virtual, module] : [module])
    : virtual;

  return {module: cast, libs: {[namespace]: bundle}, virtual: revirtual};
}

export const bundleToAttribute = (
  bundle: ShaderModule,
  name?: string,
): UniformAttribute => {
  // @ts-ignore
  const module = bundle.module ?? bundle;
  const {table: {declarations}} = module;

  const entry = bundle.entry ?? module.entry ?? name;

  for (const fn of declarations) if (fn.func) {
    const {func} = fn;
    const {type, name, parameters} = func;
    if (name === entry) {
      return {name, format: toTypeString(type), args: parameters};
    }
  }

  return {name: name ?? 'main', format: 'void', args: []};
}
