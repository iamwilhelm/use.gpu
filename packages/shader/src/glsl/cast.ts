import { UniformAttribute, ShaderModule, ParsedBundle } from './types';
import { loadVirtualModule } from './shader';
import { getProgramHash } from '../util/hash';
import { toBundle } from '../util/bundle';
import { PREFIX_CAST } from '../constants';

const NO_SYMBOLS = [] as string[];

const arg = (x: number) => String.fromCharCode(97 + x);

export const castTo = (
  bundle: ShaderModule,
  type: string,
  swizzle: string,
): ParsedBundle => {
  const {module, virtual} = toBundle(bundle);
  const {name, format, args} = bundleToAttribute(bundle);

  const {table: {hash}} = module;
  const rehash = getProgramHash(`#cast [${name} ${format}] ${hash}`);
  const symbols = ['cast', 'getValue'];
  const namespace = `${PREFIX_CAST}${rehash.slice(0, 6)}`;

  const imported = {
    at: 0,
    symbols: NO_SYMBOLS,
    name: namespace,
    imports: [{name: 'getValue', imported: name}],
  };
  const modules = [imported];

  // Code generator
  const render = (namespace: string, rename: Map<string, string>) => {
    const name = rename.get('cast') ?? 'cast';
    const accessor = rename.get('getValue') ?? 'getValue';
    return makeCastTo(name, accessor, args ?? [], format, type, swizzle);
  }

  const cast = loadVirtualModule(
    { render },
    { symbols, modules },
    'cast',
    rehash,
  );

  const revirtual = module.virtual
    ? (virtual ? [...virtual, module] : [module])
    : virtual;

  return {module: cast, libs: {[namespace]: bundle}, virtual: revirtual};
}

export const makeCastTo = (
  name: string,
  accessor: string,
  args: string[],
  from: string,
  to: string,
  swizzle: string,
) => {
  const symbols = args.map((t, i) => `${arg(i)}`);
  return `${to} ${name}(${symbols.map((s, i) => `${args[i]} ${s}`).join(', ')}) {
  return ${accessor}(${symbols.join(', ')}).${swizzle};
}
`;
}

export const bundleToAttribute = (
  bundle: ShaderModule,
  name?: string,
): UniformAttribute => {
  // @ts-ignore
  const module = bundle.module ?? bundle;
  const {table: {functions}} = module;

  const entry = bundle.entry ?? module.entry ?? name;

  for (const fn of functions) {
    const {prototype} = fn;
    if (prototype) {
      const {type, name, parameters} = prototype;
      if (name === entry) {
        return {name, format: type.name, args: parameters};
      }
    }
  }

  return {name: name ?? 'main', format: 'void', args: []};
}
