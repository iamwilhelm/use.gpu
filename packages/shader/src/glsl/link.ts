import { Tree } from '@lezer/common';
import { ShaderDefine, SymbolTable, ParsedModule, ParsedModuleCache, ParsedBundle, RefFlags as RF, DataBinding } from '../types';
import { parseShader, defineConstants, loadModule, loadModuleWithCache } from './shader';
import { rewriteUsingAST, resolveShakeOps } from './ast';
import { parseBundle, parseLinkAliases } from '../util/bundle';
import { getGraphOrder } from '../util/tree';
import { GLSL_VERSION } from '../constants';
import * as T from '../grammar/glsl.terms';
import mapValues from 'lodash/mapValues';
import { DEFAULT_CACHE } from './shader';

const TIMED = false;

const timed = (name: string, f: any) => {
  if (!TIMED) return f;
  return (...args: any[]) => {
    const t = +new Date();
    const v = f(...args);
    console.log(name, (+new Date() - t), 'ms');
    return v;
  }
}

// Override GLSL version/prefix
let PREAMBLE = `#version ${GLSL_VERSION}`;
export const setPreamble = (s: string): string => PREAMBLE = s;
export const getPreamble = (): string => PREAMBLE;

// Link a source module with static modules and dynamic links.
export const linkCode = timed('linkCode', (
  code: string,
  libraries: Record<string, string> = {},
  links: Record<string, string> = {},
  defines: Record<string, ShaderDefine> = {},
  cache: ParsedModuleCache | null = DEFAULT_CACHE,
) => {
  const main = loadModuleWithCache(code, 'main', undefined, cache);

  const parsedLibraries = mapValues(libraries, (code: string, name: string) => loadModuleWithCache(code, name, undefined, cache));
  const parsedLinkDefs = mapValues(links, (code: string, name: string) => loadModuleWithCache(code, name.split(':')[0], undefined, cache));

  return linkModule(main, parsedLibraries, parsedLinkDefs, defines);
});

// Link a bundle of parsed module + libs and dynamic links.
export const linkBundle = timed('linkBundle', (
  bundle: ParsedBundle,
  links: Record<string, ParsedBundle | ParsedModule> = {},
  defines: Record<string, ShaderDefine> = {},
) => {
  const {code} = linkBundleVirtual(bundle, links, defines);
  return code;
});

// Link a parsed module with static modules and dynamic links.
export const linkModule = timed('linkModule', (
  main: ParsedModule,
  libraries: Record<string, ParsedModule> = {},
  linkDefs: Record<string, ParsedModule> = {},
  defines: Record<string, ShaderDefine> = {},
) => {
  const {code} = linkModuleVirtual(main, libraries, linkDefs, defines);
  return code;
});

// Link a bundle of parsed module + libs, dynamic links and virtual bindings.
export const linkBundleVirtual = timed('linkBundleVirtual', (
  bundle: ParsedBundle,
  links: Record<string, ParsedBundle | ParsedModule> = {},
  defines: Record<string, ShaderDefine> = {},
) => {
  let [main, libs] = parseBundle(bundle);

  const linkDefs = mapValues(links, (bundle: ParsedBundle | ParsedModule, name: string) => {
    if ('table' in bundle) return bundle as any as ParsedModule;

    const [link, linkLibs] = parseBundle(bundle as any as ParsedBundle);
    libs = {...libs, ...linkLibs};
    return link;
  });

  return linkModuleVirtual(main, libs, linkDefs, defines);
});

// Link a parsed module with static modules, dynamic links and virtual bindings.
export const linkModuleVirtual = timed('linkModuleVirtual', (
  main: ParsedModule,
  libraries: Record<string, ParsedModule> = {},
  linkDefs: Record<string, ParsedModule> = {},
  defines: Record<string, ShaderDefine> = {},
) => {
  if (typeof main === 'string') throw new Error("Module is a string instead of an object");

  const [links, aliases] = parseLinkAliases(linkDefs);
  const {modules, exported} = loadModulesInOrder(main, libraries, links, aliases);

  const program = [getPreamble(), defineConstants(defines)] as string[];

  const hashes = new Map<string, string>();
  const namespaces = new Map<string, string>();
  const used = new Set<string>();
  const exists = new Set<string>();
  const visible = new Set<string>();
  const fixed = new Map<string, string>();

  let bindingIndex = 0;
  const nextBinding = () => bindingIndex++;
  const bindings = [] as DataBinding[][];

  for (const module of modules) {
    const {name, code, tree, table, shake, virtual} = module;
    const {hash, globals, symbols, visibles, externals, modules} = table;

    // Multiple links into same module with different name
    if (hashes.has(hash)) {
      namespaces.set(name, hashes.get(hash)!);
      continue;
    }

    // Namespace all non-global symbols outside main module
    const rename = new Map<string, string>();
    if (module !== main) {
      const namespace = reserveNamespace(module, namespaces, used);
      hashes.set(hash, namespace);

      for (const name of symbols) rename.set(name, namespace + name);
      for (const name of globals) rename.set(name, name);

      for (const name of visibles) visible.add(rename.get(name)!);
      for (const name of rename.values()) exists.add(name);

      for (const name of globals) fixed.set(namespace + name, name);
    }

    // Replace imported symbol names with target
    for (const {name: module, imports} of modules) {
      const namespace = namespaces.get(module);
      for (const {name, imported} of imports) {
        let imp = namespace + imported;
        if (fixed.has(imp)) imp = fixed.get(imp)!;
        if (!exists.has(imp)) console.warn(`Import ${name} from '${module}' does not exist`);
        else if (!visible.has(imp)) console.warn(`Import ${name} from '${module}' is private`);
        rename.set(name, imp);
      }
    }

    // Replace imported function prototype names with target
    for (const {flags, prototype} of externals) if (prototype) {
      const {name} = prototype;
      const resolved = aliases?.get(name) ?? name;
      const namespace = namespaces.get(name);

      if ((namespace === undefined) && (flags & RF.Optional)) continue;

      let imp = namespace + resolved;
      if (fixed.has(imp)) imp = fixed.get(imp)!;
      if (!exists.has(imp)) console.warn(`Link ${name}:${resolved} does not exist`);
      else if (!visible.has(imp)) console.warn(`Link ${name}:${resolved} is private`);
      rename.set(name, imp);
    }

    if (virtual) {
      // Emit virtual module in target namespace,
      // with dynamically assigned binding slots.
      const namespace = hashes.get(hash)!;
      const {code, virtuals} = virtual.render(namespace, nextBinding);

      program.push(code);
      bindings.push(...virtuals);
    }
    else {
      // Shake tree ops based on which symbols were exported
      const keep = exported.get(hash);
      const ops = shake && keep ? resolveShakeOps(shake, keep) : null;

      // Rename symbols using AST while tree shaking
      const recode = rewriteUsingAST(code, tree, rename, ops);
      program.push(recode);
    }
  }

  const code = program.join("\n");
  return {code, bindings};
});

// Load all modules references from a piece of code,
// while gathering info about what's exported (for tree shaking).
export const loadModulesInOrder = timed('loadModules', (
  main: ParsedModule,
  libraries: Record<string, ParsedModule>,
  links: Record<string, ParsedModule>,
  aliases: Map<string, string> | null,
) => {
  const graph = new Map<string, string[]>();
  const seen = new Set<string>();
  const hoist = new Set<string>();
  const exported = new Map<string, Set<string>>();

  const out = [] as ParsedModule[];

  // Traverse graph starting from main
  const queue = [{name: main.name, module: main}];
  seen.add(main.name);

  while (queue.length) {
    const next = queue.shift()!;
    const {name, module} = next;
    if (module == null) throw new Error(`Shader module ${name} undefined`);

    out.push(module);

    const {table: {modules, externals}} = module;
    const deps = [] as string[];

    // Recurse into imports
    for (const {at, name, imports} of modules) {
      const isLink = name.match(/^#/);
      const module = isLink ? links[name] : libraries[name];
      if (!module) throw new Error(`Unknown module '${name}'`);

      if (!seen.has(name)) queue.push({name, module: {...module, name}});
      seen.add(name);
      deps.push(name);

      if (at < 0) hoist.add(name);

      const {table: {hash}} = module;
      let list = exported.get(hash);
      if (!list) exported.set(hash, list = new Set());
      imports.forEach(i => list!.add(module.entry ?? i.name));
    }

    // Recurse into links
    for (const {prototype, flags} of externals) if (prototype) {
      const {name} = prototype;
      const module = links[name];
      if (!module) {
        if (flags & RF.Optional) continue;
        throw new Error(`Unlinked function '${name}'`);
      }

      if (!seen.has(name)) queue.push({name, module: {...module, name}});
      seen.add(name);
      deps.push(name);

      const {table: {hash}} = module;
      let list = exported.get(hash);
      if (!list) exported.set(hash, list = new Set());
      list.add(module.entry ?? aliases?.get(name) ?? name);
    }

    // Build module-to-module dependency graph
    graph.set(name, deps);
  }

  // Sort by graph depth
  const order = getGraphOrder(graph, main.name);
  for (let [k, v] of order.entries()) if (hoist.has(k)) order.set(k, v + 1e5);
  out.sort((a, b) => order.get(b.name)! - order.get(a.name)! || a.name.localeCompare(b.name));

  return {
    modules: out,
    exported,
  };
});

// Reserve a unique namespace based on truncated shader hash
export const reserveNamespace = (
  module: ParsedModule,
  namespaces: Map<string, string>,
  used: Set<string>,
) => {
  const {name, table: {hash}} = module;
  let namespace;
  let n = 2;

  do {
    namespace = '_' + hash.slice(0, n++) + '_';
  } while (used.has(namespace));

  namespaces.set(name, namespace);
  used.add(namespace);

  return namespace;
}
