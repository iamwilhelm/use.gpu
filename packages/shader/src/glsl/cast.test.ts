import { loadModule } from './shader';
import { linkBundle } from './link';
import { castTo } from './cast';

describe('cast', () => {
  
  it('casts', () => {
    
    const code = `
    #pragma export
    vec2 getValue() { return vec2(1.0, 2.0); }
    `;

    const mod = loadModule(code, 'code', 'getValue');
    const result = castTo(mod, 'vec3', 'xyx');
    
    const recode = linkBundle(result);
    expect(recode).toMatchSnapshot();

  });

  it('casts and links', () => {
    
    const code = `
    #pragma export
    vec2 getValue() { return vec2(1.0, 2.0); }
    `;

    const main = `
    vec3 getValue();
    void main() { getValue(); }
    `;

    const sub = loadModule(code, 'code', 'getValue');
    const mod = loadModule(main, 'main');
    const getValue = castTo(sub, 'vec3', 'xyx');

    const recode = linkBundle(mod, {getValue});
    expect(recode).toMatchSnapshot();

  });

});