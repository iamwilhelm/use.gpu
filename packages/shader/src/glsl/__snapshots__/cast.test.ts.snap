// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cast casts 1`] = `
"#version 450

    vec2 _l0_getValue() { return vec2(1.0, 2.0); }
    
vec3 cast() {
  return _l0_getValue().xyx;
}"
`;

exports[`cast casts and links 1`] = `
"#version 450

    vec2 _l0_getValue() { return vec2(1.0, 2.0); }
    
vec3 _36_cast() {
  return _l0_getValue().xyx;
}

    vec3 _36_cast();
    void main() { _36_cast(); }
    "
`;
